File: backend/app/schemas/travel.py
from pydantic import BaseModel, ConfigDict, Field, field_validator
from datetime import date, datetime
from typing import Literal

from app.utils.uploads import public_upload_name


class TravelCreate(BaseModel):
    travel_date: date
    from_address: str
    to_address: str
    total_km: float
    patient_visited: bool
    patient_name: str | None = None
    transport_mode: str | None = "Vehicle"
      # e.g., car, bus, train, etc.
    bill_amount: float | None = None    

class TravelResponse(BaseModel):
    id: int
    therapist_id: int
    therapist_name: str | None = None
    travel_date: date
    from_address: str
    to_address: str
    total_km: float
    per_km_rate: float
    travel_fare: float
    patient_visited: bool
    status: str
    claim_id: int | None = None
    patient_name: str | None = None
    transport_mode: str # e.g., car, bus, train, etc.
    bill_amount: float | None = None
    invoice_file: str | None = None
    schedule_id: int | None = None
    policy_id: int | None = None
    calculation_version: str = "decimal-v1"
    rounding_mode: str = "ROUND_HALF_UP"
    arrival_latitude: float | None = None
    arrival_longitude: float | None = None
    manual_reason: str | None = None
    manual_review_status: str | None = None
    manual_review_reason: str | None = None
    manual_revision: int = 1
    manual_review_version: int = 1
    available_actions: list[str] = Field(default_factory=list)

    @field_validator("invoice_file", mode="before")
    @classmethod
    def serialize_invoice_name(cls, value) -> str | None:
        return public_upload_name(value, "invoice")
    
    model_config = ConfigDict(from_attributes=True)

class TravelUpdate(BaseModel):
    travel_date: date
    from_address: str
    to_address: str
    total_km: float
    patient_visited: bool
    patient_name: str | None = None
    transport_mode: str | None = None # e.g., car, bus, train, etc.
    bill_amount: float | None = None
    invoice_file: str | None = None


class ManualTravelDecision(BaseModel):
    decision: Literal["approved", "changes_requested"]
    reason: str = Field(min_length=5, max_length=500)
    version: int = Field(ge=1)


class ManualTravelReviewEventResponse(BaseModel):
    id: int
    event_type: str
    actor_id: int
    actor_name: str | None = None
    from_status: str | None = None
    to_status: str
    reason: str
    revision: int
    created_at: datetime

File: mobile/src/types/travel.ts
export interface TravelResponse {
  id: number;
  therapist_id: number;
  therapist_name: string | null;
  travel_date: string;
  from_address: string;
  to_address: string;
  total_km: number;
  per_km_rate: number;
  travel_fare: number;
  patient_visited: boolean;
  status: string;
  claim_id: number | null;
  patient_name: string | null;
  transport_mode: string;
  bill_amount: number | null;
  invoice_file: string | null;
  schedule_id: number | null;
  arrival_latitude: number | null;
  arrival_longitude: number | null;
  manual_reason: string | null;
  manual_review_status: string | null;
  manual_review_reason: string | null;
  manual_revision: number;
  manual_review_version: number;
  available_actions: string[];
}

export interface ManualTravelReviewEvent {
  actor_id: number;
  actor_name: string | null;
  created_at: string;
  event_type: string;
  from_status: string | null;
  id: number;
  reason: string;
  revision: number;
  to_status: string;
}
